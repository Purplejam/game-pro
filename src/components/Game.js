import React from 'react';
import {motion, AnimateSharedLayout} from 'framer-motion';
import styled from 'styled-components';
import {gameDetailAction} from '../actions/gameDetailAction.js';
import {useDispatch, useSelector} from 'react-redux';
import { Link } from "react-router-dom";
import default_img from "../img/default_img.png";
import {fadeIn, item} from '../animation.js';

//Styles
const StyledGame = styled(motion.div)`
	img {
		width: 100%;
		height: 40vh;
		object-fit: cover;
	}
	h3 {
		cursor: pointer;
	}
	min-height: 30vh;
	box-shadow: 0px 5px 30px rgba(0, 0, 0, 0.2);
	text-align: center;
	border-radius: 1rem;
	overflow: hidden; 
`
//Media resize
export const smallImage = (imagePath, size) => {
  const image = imagePath.match(/media\/screenshots/)
    ? imagePath.replace(
        "media/screenshots",
        `media/resize/${size}/-/screenshots`
      )
    : imagePath.replace("/media/games/", `/media/resize/${size}/-/games/`);
  return image;
};


function Game({name, released, image, id, slug}) {
	const dispatch = useDispatch();
	const layoutId = id.toString();

	function idGameHandler() {
		document.body.style.overflow = 'hidden';
		dispatch(gameDetailAction(id, slug));
	}

	return(
		<StyledGame variants={item} onClick={idGameHandler}>
			<Link to={`/game/${id}`}>
				<motion.h3>{name}</motion.h3>	
				<p>{released}</p>
				<motion.img src={image == null ? default_img : image}/>
			</Link>	
		</StyledGame>
		)
}

export default Game;