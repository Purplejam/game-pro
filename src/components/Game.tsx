import {motion} from 'framer-motion';
import styled from 'styled-components';
import {gameDetailAction} from '../actions/gameDetailAction';
import {useDispatch} from 'react-redux';
import { Link } from "react-router-dom";
import default_img from "../img/default_img.png";
import { item } from '../animation';
import star from '../img/star-full.png';
import { gameType } from './Home';
import { AppStateType } from '../reducers/index';
import { ThunkDispatch } from 'redux-thunk';
import { Action } from 'redux';


//styles
const StyledGame = styled(motion.div)`
	img {
		width: 100%;
		height: 40vh;
		object-fit: cover;
	}
	h3 {
		cursor: pointer;
		padding: 2rem 0rem 0.5rem 0rem;
	}
	min-height: 30vh;
	box-shadow: 0px 5px 30px rgba(0, 0, 0, 0.2);
	text-align: center;
	border-radius: 1rem;
	overflow: hidden; 
	.game-meta-info {
		display: flex;
		flex-direction: row;
		justify-content: space-between;
		align-content: space-between;
		padding: 1rem 2rem;
			img {
				width: 1.2rem;
				height: 1.2rem;
				display: inline-block;
				margin-left: 0.5rem;
				margin-top: -0.1rem;
				vertical-align: middle;
			}
			div, p {
				display: inline-block;
				font-weight: 700;
			}
			.game-raiting {
				display: flex;
				align-items: center;
			}
	}
	.game-genres {
		text-align: center;
		padding: 1rem 2rem;
	}
 @media (max-width: 768px) {
 }
`
//media resize
export const smallImage = (imagePath: string, size: number) => {
  const image = imagePath.match(/media\/screenshots/)
    ? imagePath.replace(
        "media/screenshots",
        `media/resize/${size}/-/screenshots`
      )
    : imagePath.replace("/media/games/", `/media/resize/${size}/-/games/`);
  return image;
};

//rating color
const ratingColor = (rate: number) => {
	if (rate >= 70 && rate < 100) {
		return 'green';
	} else if (rate >= 50 && rate < 70) {
		return 'orange';
	} else if (rate > 0 && rate < 50) {
		return 'red';
	}
}


function Game({name, released, background_image, id, slug, genres, ratings, added}: gameType) {
	const dispatch: ThunkDispatch<AppStateType, void, Action> = useDispatch();

	function idGameHandler() {
		document.body.style.overflow = 'hidden';
		dispatch(gameDetailAction(id, slug));
	}

	return(
		//@ts-ignore
		<StyledGame variants={item} onClick={idGameHandler}>
			<Link to={`/game/${id}`}>
				<motion.h3>{name}</motion.h3>	
				<p>Released: {released}</p>
				<motion.img src={background_image == null ? default_img : smallImage(background_image, 420)}/>
				{(genres.length > 0) 
					? <p className="game-genres">Genre{genres.length > 1 ? 's' : null}: {genres.map(genre => <span key={genre.id}>{genre.name + ' '}</span>)}</p>
					: <p className="game-genres">Genre: video game</p>}
				<div className="game-meta-info">
					{ratings.length !== 0 ? <p style={{color: ratingColor(ratings[0].percent)}}>Raiting: {Math.floor(ratings[0].percent)}</p> : null}
					<div className="game-raiting">
						<p>{added}</p>
						<img src={star} alt=""/>
					</div>
				</div>
			</Link>	
		</StyledGame>
		)
}

export default Game;