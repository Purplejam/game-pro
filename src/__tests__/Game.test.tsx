import {render, screen} from '@testing-library/react';
import Game from '../components/Game';
import {store} from '../store/store';
import { Provider } from 'react-redux';
import React from 'react';
import { BrowserRouter} from "react-router-dom";
import '@testing-library/jest-dom';

 
const mockGame = {
	id: 1234,
	name: 'GameTest',
	ratings: [{percent: 12}],			
	slug: 'game-test',
	added: 202020,
	released: '22-12-2022',
	background_image: 'string',
	genres: [{
			name: 'RPG',
			id: 12
		}]
}

describe('<Game/>', () => {
	it('renders properly the game component', () => {
		const {container} = render(
			<BrowserRouter>
				<React.StrictMode>
					<Provider store={store}>
						 <Game {...mockGame}/>
					</Provider>
				</React.StrictMode>
			</BrowserRouter>
			)

		const nameText = screen.getByText('GameTest');
		expect(nameText).toBeInTheDocument();

		const genreText = screen.getByText('RPG');
		expect(genreText).toBeInTheDocument();

		const addedText = screen.getByText('202020');
		expect(addedText).toBeInTheDocument();

		const releasedText = screen.getByText('Released: 22-12-2022');
		expect(releasedText).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	})
})

