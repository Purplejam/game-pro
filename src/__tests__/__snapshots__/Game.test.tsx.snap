// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game/> renders properly the game component 1`] = `
<div>
  <div
    class="sc-bczRLJ eymvcN"
  >
    <a
      href="/game/1234"
    >
      <h3>
        GameTest
      </h3>
      <p>
        Released: 
        22-12-2022
      </p>
      <img
        src="string"
      />
      <p
        class="game-genres"
      >
        Genre
        : 
        <span>
          RPG 
        </span>
      </p>
      <div
        class="game-meta-info"
      >
        <p
          style="color: red;"
        >
          Raiting: 
          12
        </p>
        <div
          class="game-raiting"
        >
          <p>
            202020
          </p>
          <img
            alt=""
            src="star-full.png"
          />
        </div>
      </div>
    </a>
  </div>
</div>
`;
